prosemirror/model/from_dom.py

```diff
@@ -0,0 +1,1072 @@
+				content_dom = rule.content_element(dom_)
+			elif rule.content_element is not None:
+				content_dom = rule.content_element
+
+			self.find_around(dom_, content_dom, True)
+			self.add_all(content_dom)
+
+		if sync and self.sync(start_in):
+			self.open -= 1
+
+		if mark is not None:
+			self.remove_pending_mark(mark, start_in)
+
+	def add_all(self, parent: DOMNode, start_index: int | None = None, end_index: int | None = None) -> None:
+		index = start_index if start_index is not None else 0
+
+		try:
+			dom_ = list(parent)[index]
+		except IndexError:
+			pass
+		else:
+			end = None if end_index is None else list(parent)[end_index]
+
+			while dom_ != end:
+				self.find_at_point(parent, index)
+				self.add_DOM(dom_)
+
+				dom_ = dom_.getnext()
+				index += 1
+
+		self.find_at_point(parent, index)
+
+	def find_place(self, node: Node) -> bool:
+		route: List[NodeType] | None = None
+		sync: NodeContext | None = None
+
+		depth = self.open
+		while depth >= 0:
+			cx = self.nodes[depth]
+			found = cx.find_wrapping(node)
+			if found is not None and (route is None or len(route) > len(found)):
+				route = found
+				sync = cx
+
+				if found is None:
+					break
+
+			if cx.solid:
+				break
+
+			depth -= 1
+
+		if route is None:
+			return False
+
+		if sync is not None:
+			self.sync(sync)
+
+		for r in route:
+			self.enter_inner(r, None, False)
+
+		return True
+
+	def insert_node(self, node: Node) -> bool:
+		if node.is_inline and self.needs_block and self.top.type is None:
+			block = self.textblock_from_context()
+			if block is not None:
+				self.enter_inner(block)
+
+		if self.find_place(node):
+			self.close_extra()
+
+			top = self.top
+			top.apply_pending(node.type)
+
+			if top.match is not None:
+				top.match = top.match.match_type(node.type)
+
+			marks = top.active_marks
+			for mark in node.marks:
+				if top.type is None or top.type.allows_mark_type(mark.type):
+					marks = mark.add_to_set(marks)
+
+			top.content.append(node.mark(marks))
+
+			return True
+
+		return False
+
+	def enter(self, type_: NodeType, attrs: Attrs | None = None, preserve_ws: WSType = None) -> bool:
+		ok = self.find_place(type_.create(attrs))
+		if ok:
+			self.enter_inner(type_, attrs, True, preserve_ws)
+
+		return ok
+
+	def enter_inner(self, type_: NodeType, attrs: Attrs | None = None, solid: bool = False, preserve_ws: WSType = None) -> None:
+		self.close_extra()
+
+		top = self.top
+		top.apply_pending(type_)
+
+		if top.match is not None:
+			top.match = top.match.match_type(type_)
+
+		options = ws_options_for(type_, preserve_ws, top.options)
+
+		if (top.options & OPT_OPEN_LEFT) and len(top.content) == 0:
+			options |= OPT_OPEN_LEFT
+
+		self.nodes.append(
+			NodeContext(type_, attrs, top.active_marks, top.pending_marks, solid, None, options)
+		)
+
+		self.open += 1
+
+	def close_extra(self, open_end: bool = False) -> None:
+		i = len(self.nodes) - 1
+
+		if i > self.open:
+			while i > self.open:
```

Comment:
This loop as well I would use for with range to avoid having to update the counter at the end. I won't leave more comments about this, but there are a few more cases where I feel the same way in the rest of the file